compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -838,6 +856,119 @@ runScriptsRule opts =
       let (diags, results) = unzip scenarioResults
       pure (catMaybes diags, Just results)
 
+runScriptsPkg ::
+       NormalizedFilePath
+    -> LF.ExternalPackage
+    -> [LF.ExternalPackage]
+    -> Action ([FileDiagnostic], Maybe [(VirtualResource, Either SS.Error SS.ScenarioResult)])
+runScriptsPkg projRoot extPkg pkgs = do
+    Just scenarioService <- envScenarioService <$> getDamlServiceEnv
+    ctx <- contextForPackage projRoot pkg
+    ctxIdOrErr <- liftIO $ SS.getNewCtx scenarioService ctx
+    ctxId <-
+        liftIO $
+        either
+            (throwIO . ScenarioBackendException "Failed to create scenario context")
+            pure
+            ctxIdOrErr
+    scenarioContextsVar <- envScenarioContexts <$> getDamlServiceEnv
+    liftIO $ modifyMVar_ scenarioContextsVar $ pure . HashMap.insert projRoot ctxId
+    scenarioResults <-
+        liftIO $
+        forM scenarios $ \scenario -> do
+            (vr, res) <-
+                runScript
+                    scenarioService
+                    pkgName'
+                    ctxId
+                    scenario
+            pure (toDiagnostic res, (vr, res))
+    let (diags, results) = unzip scenarioResults
+    pure (catMaybes diags, Just results)
+  where
+    pkg = LF.extPackagePkg extPkg
+    pkgId = LF.extPackageId extPkg
+    pkgName' =
+        toNormalizedFilePath' $
+        T.unpack $
+        maybe (LF.unPackageId pkgId) (LF.unPackageName . LF.packageName) $ LF.packageMetadata pkg
+    world = LF.initWorldSelf pkgs pkg
+    scenarios =
+        map fst $
+        concat
+            [ scriptsInModule (EnableScripts True) mod
+            | mod <- NM.elems $ LF.packageModules pkg
+            , LF.moduleName mod /= LF.ModuleName ["Daml", "Script"]
+            ]
+    toDiagnostic :: Either SS.Error SS.ScenarioResult -> Maybe FileDiagnostic
```

Comment:
this looks duplicated with what we do for scenarios/scripts in the current package. Probably worth factoring this and perhaps some of the other logic in this function.

Reply:
factored this out.